examples/pruning/lightgbm_with_custom_cv_function.py

```diff
@@ -0,0 +1,404 @@
+    USE_STOPPING = True
+
+    # fix number of boosting rounds
+    # note: will be overridden if USE_STOPPING is True
+    FIXED_ROUNDS = -1
+    if not USE_STOPPING:
+        FIXED_ROUNDS = 500
+
+    # set model objective
+    OBJECTIVE = "binary"
+    # OBJECTIVE = "regression"
+
+    # set metric for model training and validation
+    # as well as its optimization direction
+    if OBJECTIVE in ["binary", "classification"]:
+        METRIC = "auc"
+        OPT_DIR = "maximize"
+
+    elif OBJECTIVE in ["regression"]:
+        METRIC = "rmse"
+        OPT_DIR = "minimize"
+
+    # whether to make Optuna experiment reproducible
+    # (with a repeatable set of metrics for each set of seeds)
+    # at the cost of performance (single worker)
+    MAKE_REPRODUCIBLE = True
+
+    # define the number of threads
+    # to be used by lightgbm
+    # in each of the workers
+    LGBM_NUM_THREADS = 2
+
+    # set number of Optuna workers
+    WORKERS_NUM = 2
+
+    if MAKE_REPRODUCIBLE:
+        # turn off multiprocessing in Optuna samplers to ensure reproducible results
+        # (https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-obtain-reproducible-optimization-results)
+        WORKERS_NUM = 1
+
+    # fix seed for data partitioning (train_test_split())
+    # and morel training (lgbm.fit())
+    SEED = 123
+
+    # select sampler for Optuna:
+    # (see more samplers at: https://optuna.readthedocs.io/en/stable/reference/samplers.html)
+    # - TPE (Tree-structured Parzen Estimator) sampling algo;
+    # note we fix sampler seed to make the sampling process deterministic
+    SAMPLER_SEED = 123
+    OPTUNA_SAMPLER = TPESampler(seed=SAMPLER_SEED)
+
+    # whether to prune unpromising trials to speed up studies
+    USE_PRUNING = True
+
+    # select pruner for Optuna:
+    # (see more pruners at: https://optuna.readthedocs.io/en/latest/reference/pruners.html)
+    # - median pruner
+    OPTUNA_PRUNER = MedianPruner()
+
+    # set verbosity level for lightgbm
+    VERBOSE_LGBM = -1
+
+    # Experiment
+
+    # Load data
+
+    train_x_df, _, train_y_df, _ = load_sklearn_toy_data(
+        objective=OBJECTIVE, test_size=0.01, as_df=True, seed=SEED
+    )
+
+    # Optimize params using Optuna
+
+    optuna_best_metrics = {}
+    optuna_best_params = {}
+
+    # Instantiate the custom function
+    objective = Objective_custom(
+        train_x_df,
+        train_y_df,
+        objective=OBJECTIVE,
+        eval_metric=METRIC,
+        eval_name=CV_EVAL_NAME,
+        folds=FOLDS,
+        fixed_rounds=FIXED_ROUNDS,
+        use_stopping=USE_STOPPING,
+        use_pruning=USE_PRUNING,
+        n_jobs=LGBM_NUM_THREADS,
+        seed=SEED,
+        verbosity=VERBOSE_LGBM,
+    )
+
+    study = optuna.create_study(sampler=OPTUNA_SAMPLER, pruner=OPTUNA_PRUNER, direction=OPT_DIR)
+
+    start_time = datetime.now()
+
+    # run Optuna optimization over the specified number of trials
+    study.optimize(objective, n_trials=TRIALS, n_jobs=WORKERS_NUM)
```

Comment:
IMO, `n_jobs>1` will hardly improve the performance due to the GIL. How about stop using it?
Even so, `n_jobs` of LightGBM will utilize the multi-core processors effectively.

```suggestion
    study.optimize(objective, n_trials=TRIALS)
```

Reply:
Here I have to kindly disagree:) From my experience on multi-CPU machines (in docker containers), muli-threading built into LGBM has its limits (notably, by Guolin Ke's own admission in one discussion on Githhub) LGBM internal parallelism does not work beyond certain point (specifically, when the user tries to access threads from more than 1 physical CPU socket). In practice, the optimum is around 8-16 threads, depending on data shape. You can see some of my public results trying to establish such optimum [here](https://github.com/mirekphd/gbdt-algos-performance-tests/blob/master/20200812/analyze-results-of-gpu-vs-cpu-tests-multiple-versions-20200812.ipynb).

Enter multi-processing (tested either with `loky` or `multiprocessing` backend, as `dask` does not work), which allows us to bypass this limitation of multithreading built into LightGBM, because processes can be run on separate CPU sockets. If you wish, I can supply results of additional performance tests of this PR script, e.g. time taken on average by an Optuna trial) for the multiprocessing vs. multithreading case?

Reply:
The most likely reason why multiprocessing did not seem to work "out-of-the-box" was this reproducibility enforcement in my example attained by... switching off multiprocessing by setting `WORKERS_NUM` back from 2 to 1 (in line [#L311](https://github.com/mirekphd/optuna/blob/lgbm-w-cust-cv-fun/examples/pruning/lightgbm_with_custom_cv_function.py#L311)).

We might change the current setting of `MAKE_REPRODUCIBLE` from `True` to `False` or set `WORKERS_NUM` to 1 instead of 2 (to require action that makes it more likely to notice the role of `MAKE_REPRODUCIBLE`) if you think others will also make the same conclusions about GIL being the culprit here, rather than the desire to obtain the same result at each Optuna run.

Reply:
Thank you for your detailed comments. 

> `MAKE_REPRODUCIBLE `

I agree. To make the optimization reproducible, we need to set `n_jobs=1`. So, I think it's reasonable to set `WORKERS_NUM` to `1` by default.

>  switching off multiprocessing by setting `WORKERS_NUM` back from 2 to 1

A nit, but Optuna utilizes the multi-threading, not multiprocessing when we set `n_jobs>1`.